File: C++_code/Lecture_25.cpp
                                                //Lecture_25
                                                //July-21

#include<iostream>
#include<string>
#include<vector>
#include<queue>
#include<stack>
using namespace std;


// PRE-REQUIST for graph

class Edge                                   
{                                               //HERE, we are just going to create
    public:                                     //a graph data by using class and 
        int nebr;                               //a vector of vector of class.
        int wt; 
};

vector<vector<Edge>> graph;                      //DATA TYPE TO REPRESENT GRAPH.


void addEdge(int v1,int v2, int wt)                    //it will add the edge in the graph data
{                                                    
    Edge e1;
    e1.nebr=v2;
    e1.wt=wt;
    graph[v1].push_back(e1);

    Edge e2;
    e2.nebr=v1;
    e2.wt=wt;
    graph[v2].push_back(e2);    
}

//REVIEWED.







//1. PRIMS_ALGORITHM     (to get minimum spanning tree)

void addEdge(vector<vector<Edge>>& mst,int v,int av,int c)            //PRIMS ALGORITHM.
{
    Edge e1;
    e1.nebr=av;
    e1.wt=c;
    mst[v].push_back(e1);
    
    Edge e2;
    e2.nebr=v;
    e2.wt=c;
    mst[av].push_back(e2);
}

class pPair                //primes pair class.
{
    public:
    int vertex;
    int aVertex;
    int cost;

    pPair(int v,int av,int c)
    {
        this->vertex=v;
        this->aVertex=av;
        this->cost=c;
    }

    pPair()
    {
        this->vertex=0;
        this->aVertex=0;
        this->cost=0;
    }

    bool operator>(const pPair& other) const
    {
        return this->cost > other.cost;
    }

};

void display(vector<vector<Edge>> mst)            //to display a specific graph.                                          //it will display all the graph data
{
    for(int i=0;i<mst.size();i++)
    {
        cout<<i<<" ->";
        for(int j=0;j<mst[i].size();j++)
        {
            Edge n=mst[i][j];
            cout<<"["<<n.nebr<<"-"<<n.wt<<"]";
        }
        cout<<endl;
    }
}


void prims()                                      //prims algorithm.(to find 
{                                                 //minimum spanning tree)
    vector<bool> visited (graph.size(), false);
    vector<vector<Edge>> mst (graph.size(),vector<Edge>());
    pPair pp;
    pp.vertex=0;
    pp.aVertex=-1;
    pp.cost=0;

    priority_queue<pPair, vector<pPair>, greater<pPair>> q;
    q.push(pp);

    while(q.size()>0)
    {
        pPair rem=q.top();
        q.pop();

        if(visited[rem.vertex]==true)
        {
            continue;
        }
        visited[rem.vertex]=true;

        if(rem.aVertex!=-1)
            addEdge(mst,rem.vertex,rem.aVertex,rem.cost);

        for(int i=0;i<graph[rem.vertex].size();i++)
        {
            Edge n=graph[rem.vertex][i];
            if(visited[n.nebr]==false)
            {
                pPair p(n.nebr,rem.vertex,n.wt);
                q.push(p);
            }
        }
    }

    display(mst);
}


//REVIEWED.








//2. TOPOLOGICAL_SORT

vector<vector<Edge>> topoGraph;                     //graph for topological sort.

void addEdgeDirected(int s,int r,int c)        //to add edges in directed graph. 
{
    Edge e1;
    e1.nebr=r;
    e1.wt=c;
    topoGraph[s].push_back(e1);
}



void topologicalCompo(int i,vector<bool> &visited,stack<int> &s)       //this is similar to gscc. 
{                            
    visited[i] = true;                                          
    for(int j=0;j<topoGraph[i].size();j++)
    {
        Edge n=topoGraph[i][j];
        if(visited[n.nebr]==false)
        {   
            topologicalCompo(n.nebr,visited,s);
        }
    }

    s.push(i);
}



void topological()                               //this is similar to gcc
{                                                
    stack<int> s;
    vector<bool> visited (topoGraph.size(),false);
    
    for(int i=0;i<topoGraph.size();i++)
    {
        if(visited[i]==false)
        {
            topologicalCompo(i,visited,s);
        }
    }

    while(s.size()!=0)               //printing the topo sort order, or reverse order of stack.
    {   
        cout<<s.top()<<" ";
        s.pop();
    }

}


//reviewed.









int main()
{
    //PRE_REQUIST for graph

    graph.push_back(vector<Edge>()); //0
    graph.push_back(vector<Edge>()); //1
    graph.push_back(vector<Edge>()); //2
    graph.push_back(vector<Edge>()); //3
    graph.push_back(vector<Edge>()); //4
    graph.push_back(vector<Edge>()); //5
    graph.push_back(vector<Edge>()); //6

    addEdge(0,1,10);
    addEdge(1,2,10);
    addEdge(2,3,10);
    addEdge(0,3,40);
    addEdge(3,4,2);
    addEdge(4,5,3);
    addEdge(5,6,3);
    addEdge(4,6,8);

    vector<bool> visited (graph.size(), false);

    //reviewed.
    





    //1. PRIMS_ALGORITHM   (to get the minimum spanning tree)

    /*
    prims();
    */

    //reviewed.






    //2.TOPOLOGICAL_SORT
    
    /*
    topoGraph.push_back(vector<Edge>());   //0
    topoGraph.push_back(vector<Edge>());   //1
    topoGraph.push_back(vector<Edge>());   //2
    topoGraph.push_back(vector<Edge>());   //3
    topoGraph.push_back(vector<Edge>());   //4
    topoGraph.push_back(vector<Edge>());   //5
    topoGraph.push_back(vector<Edge>());   //6

    addEdgeDirected(0,1,1);
    addEdgeDirected(1,2,1);
    addEdgeDirected(2,3,1);
    addEdgeDirected(0,4,1);
    addEdgeDirected(5,4,1);
    addEdgeDirected(4,3,1);
    addEdgeDirected(5,6,1);
    addEdgeDirected(6,3,1);

    topological();
    */

    //reviewed.

}

File: JAVA_code/Lecture_25.java
                                                    //Lecture_25
                                                    //July-21

import java.util.*;
public class Lecture_25
{   

    //Pre-requist for dealing with graph.

    public static class Edge {
        int nebr = -1;
        int wt = -1;

        Edge(int n, int w) {
            this.nebr = n;
            this.wt = w;
        }
    }
    static ArrayList<ArrayList<Edge> > graph = new ArrayList<>();

    
    static void addEdge(int u, int v, int wt) {
        graph.get(u).add(new Edge(v, wt));
        graph.get(v).add(new Edge(u, wt));
    }


    static void display() {
        for(int i = 0; i < graph.size(); i++) {
            System.out.print(i + " -> ");
            for(int j = 0; j < graph.get(i).size(); j++) {
                System.out.print( "[ " + graph.get(i).get(j).nebr + ", " + graph.get(i).get(j).wt + "], ");
            }
            System.out.println();
        }
    }

    //REVIEWED.







    //1. PRIMS_ALGORITHM        (to get minimum spanning tree)

    static class PrimsPair implements Comparable<PrimsPair>{
        int vertex = -1;
        int cost = -1;
        int previousVertex = -1;

        PrimsPair(int v, int pv, int c) {
            this.vertex = v;
            this.cost = c;
            this.previousVertex = pv;
        }

        public int compareTo(PrimsPair other) {
            return this.cost - other.cost;
        }
    }

    
    static ArrayList<ArrayList<Edge>> mst = new ArrayList<>();      //final minimum spanning tree.

    static void addEdge(ArrayList<ArrayList<Edge>> G, int u, int v, int c) {    //add-edge for any graph G
        G.get(u).add(new Edge(v,c));
        G.get(v).add(new Edge(u,c));
    }



    static void display(ArrayList<ArrayList<Edge>> G) {         //display for any graph G.
        for(int i = 0; i < G.size(); i++) {
            System.out.print(i + " -> ");
            for(int j = 0; j < G.get(i).size(); j++) {
                Edge e = G.get(i).get(j);
                System.out.print("[" + e.nebr + ", " + e.wt + "] ");
            }
            System.out.println();
        }
    }



    static void prims() {
        PriorityQueue<PrimsPair> pq = new PriorityQueue<>();
        boolean[] visited = new boolean[graph.size()];
        PrimsPair p = new PrimsPair(0, -1, 0);
        pq.add(p);

        while(pq.isEmpty() == false) {
            PrimsPair rem = pq.poll();     //GET and REMOVE

            if(visited[rem.vertex] == true)   //if already visited, thn continue because
                continue;                     //that vertex would already be added in the mst.
            
            visited[rem.vertex] = true;        //MARK

            if(rem.previousVertex != -1) {       //WORK
                addEdge(mst, rem.vertex, rem.previousVertex, rem.cost);
            }
            
            for(int i = 0; i < graph.get(rem.vertex).size(); i++) {       //ADD NEBR
                Edge e = graph.get(rem.vertex).get(i);
                if(visited[e.nebr] == false) {
                    pq.add(new PrimsPair(e.nebr, rem.vertex, e.wt));
                }
            }
        }

        display(mst);               //displaying the mst we just created,for given graph.
    }

    //REVIEWED.








    //2. TOPOLOGICAL_SORT      

    static void addEdgeDirected(ArrayList<ArrayList<Edge>> dag, int u, int v, int c) {
        dag.get(u).add(new Edge(v, c));
    }



    //this is component function like gscc.
    public static void topoSortCompo(ArrayList<ArrayList<Edge>> dag, Stack<Integer> st, boolean[] visited, int v) {
        visited[v] = true;                  //marking visited.

        for(int i = 0; i < dag.get(v).size(); i++) {        //getting the reachable nebrs first.
            Edge e = dag.get(v).get(i);
            if(visited[e.nebr] == false)
                topoSortCompo(dag, st, visited, e.nebr);
        }

        st.push(v);                                     //inserting the vertex in post when all,
    }                                                   //its possible reachable nebrs are already pushed in stack.



    public static void topoSort(ArrayList<ArrayList<Edge>> dag) {       //driving function like gcc.
        
        boolean[] visited = new boolean[dag.size()];
        Stack<Integer> st = new Stack<>();
        
        for(int v = 0; v < dag.size(); v++) {
            if(visited[v] == false)
                topoSortCompo(dag, st, visited, v);
        }

        while(st.size() != 0) {                         //printing the reverse order of stack,
            System.out.print(st.peek() + " ");          //or we can say correct topological sort order.
            st.pop();
        }
    }


    //REVIEWED.





    public static void main(String[] args)
    {
        //pre-requist for graph
        for(int i = 0; i < 7; i++) {
            graph.add(new ArrayList<>());
        }
        
        addEdge(0,1,10);
        addEdge(1,2,10);
        addEdge(2,3,10);
        addEdge(0,3,40);
        // addEdge(2,5,5);            //extra edge between 2,5.
        addEdge(3,4,2);
        addEdge(4,5,3);
        addEdge(5,6,3);
        addEdge(4,6,8);
    
        boolean[] visited = new boolean[graph.size()];
        //reviewed.
    
    
        

        //1. PRIMS_ALGORITHM      (to create minimum spanning tree)
        
        /*       
        for(int i = 0; i < graph.size(); i++) {
            mst.add(new ArrayList<Edge>());
        }
        prims();
        */

        //reviewed.





        //2. TOPOLOGICAL_SORT           
        
        /*
        ArrayList<ArrayList<Edge>> dag = new ArrayList<>();
        for(int i = 0; i < 7; i++) {
            dag.add(new ArrayList<Edge>());
        }
        addEdgeDirected(dag, 0, 1, 1);
        addEdgeDirected(dag, 0, 4, 1);
        addEdgeDirected(dag, 1, 2, 1);
        addEdgeDirected(dag, 2, 3, 1);
        addEdgeDirected(dag, 4, 3, 1);
        addEdgeDirected(dag, 5, 4, 1);
        addEdgeDirected(dag, 5, 6, 1);
        addEdgeDirected(dag, 6, 3, 1);

        topoSort(dag);
        */

        //reviewed.


    }

}